Add ERC: Dual-Mode Fungible Tokens

0xRowan opened the pull request:
This EIP defines a minimal interface standard for privacy tokens on Ethereum.

While developing privacy solutions for the Ethereum ecosystem—including wrapper protocols (converting [ERC-20](./eip-20) ↔ privacy tokens) and dual-mode tokens (combining public and private balances)—we identified a recurring need for standardized privacy primitives. Without a common interface, each implementation reinvents commitments, nullifiers, and note encryption, leading to ecosystem fragmentation.

This standard provides that common foundation. It enables:
- **Wrapper protocols**: Implement this interface for their privacy layer
- **Dual-mode tokens protocols**: Combine standards via `contract DMT is ERC20, IZRC20`

By unifying the privacy token interface, we facilitate the development of wrapper and dual-mode protocols, accelerating Ethereum's privacy ecosystem growth.

Discussion

Community discussion: 
https://ethereum-magicians.org/t/erc-8086-privacy-token/26623
https://ethereum-magicians.org/t/erc-8085-dual-mode-fungible-tokens/26592

0xRowan committed:
Add ERC-8099: Native Privacy Token Standard

0xRowan committed:
remove 'Standard'

0xRowan committed:
Update erc-8099.md

abcoathup commented on ERCS/erc-8099.md:
```diff
@@ -0,0 +1,482 @@
+title: Native Privacy Token Interface
+description: A minimal interface for native privacy-preserving fungible tokens on Ethereum
+author: Rowan (@0xRowan)
+discussions-to: https://ethereum-magicians.org/t/draft-native-privacy-token-interface/26623
```
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-8086-native-privacy-token-interface/26623
```
Updated with assigned number

0xRowan committed:
Rename ERC-8099 to ERC-8086 per editor assignment

- Update EIP number from 8099 to 8086
- Simplify title to "Privacy Token"
- Update discussions-to URL

0xRowan committed:
Interface Update: Nullifier Query Function

0xRowan committed:
Add Reference Implementation

0xRowan committed:
Remove external links

0xRowan committed:
make nullifiers and activeSubtreeRoot required in IZRC20 interface

0xRowan committed:
Add ERC-8085: Dual-Mode Fungible Tokens (depends on ERC-8086)

0xRowan committed:
Reference Implementation

ERC-8085 Reference Implementation
ERC-8086 Reference Implementation

0xRowan committed:
fix: broken internal links in 8085/8086 to pass htmlproofer

0xRowan committed:
Refactored the reference implementation

0xRowan committed:
Add permissionless note

natalejoe reviewed (approved):
13591e76ef268fb05d2932eb199a237e4726d22c4f64https://github.com/ethereum/ERCs/pull/1359https://ethereum-magicians.org/t/eip-7928-breakout-9-dec-17-2025/27078?u=natalejoeFixes #1398

0xRowan commented:
> The test app appears to be down: https://testdmt.zkprotocol.xyz/
> 
> Is there a new deployment or URL?

You can try again — it’s working properly now.

sondotpin commented:
thanks. Another question, what's different between your protocol with Railgun

0xRowan committed:
Add Privacy Configuration File mechanism for client interoperability

This enables universal client support by allowing tokens to declare their
  implementation-specific parameters via an off-chain JSON configuration:
  - Proof system (protocol, curve, fieldSize)
  - Circuit artifacts URLs and public signals schemas
  - Note encryption algorithm and curve parameters
  - Merkle tree configuration
  - Optional service endpoints

0xRowan committed:
Update erc-8086.md

0xRowan committed:
Update erc-8086.md

0xRowan committed:
Update erc-8086.md

SamWilsn committed:
Update erc-8085.md

SamWilsn committed:
Update erc-8085.md

SamWilsn commented on ERCS/erc-8085.md:
```diff
@@ -0,0 +1,457 @@
+   - Can respond to changing jurisdictional rules by adjusting mode usage
+
+
+### BURN_ADDRESS Requirement for toPublic
```
```suggestion
### `BURN_ADDRESS` Requirement for `toPublic`
```

Make sure all inline code instances are in backticks.

SamWilsn commented on ERCS/erc-8085.md:
```diff
@@ -0,0 +1,457 @@
+
+- All ERC-20 functions operate on transparent balances
+- All standard ERC-20 events are emitted
+- All ERC-8086 (IZRC20) functions and events are supported for privacy mode
```
```suggestion
- All ERC-8086 (`IZRC20`) functions and events are supported for privacy mode
```

SamWilsn commented on ERCS/erc-8085.md:
````diff
@@ -0,0 +1,457 @@
+   - Note B (50) sent to attacker's privacy address ← Still spendable in privacy mode!
+   - Note C (50 change)
+   Result: 50 + 50 + 50 = 150 (created 50 out of thin air!)
+```
````
````suggestion
1. Hold Note A (100 privacy balance)
2. Call toPublic(50) with proof sending output to attacker's own privacy address
3. If contract skips BURN_ADDRESS check:
   - Attacker receives 50 public balance (converted from privacy mode)
   - Note B (50) sent to attacker's privacy address ← Still spendable in privacy mode!
   - Note C (50 change)
   Result: 50 + 50 + 50 = 150 (created 50 out of thin air!)
````

This isn't code.

SamWilsn commented on ERCS/erc-8085.md:
````diff
@@ -0,0 +1,457 @@
+   Result: 50 + 50 + 50 = 150 (created 50 out of thin air!)
+```
+
+**Mitigation**: Implementations MUST ensure the converted value cannot be spent in privacy mode. For BURN_ADDRESS approach:
````
Requirements, even security ones, should be in the Specification section, though you should highlight and discuss them in this section.

SamWilsn commented on assets/erc-8085/README.md:
```diff
@@ -0,0 +1,180 @@
+5. Convert to privacy mode: `toPrivacy()`
+6. Perform privacy transfers
+7. Convert back to public: `toPublic()`
+8. Verify transactions on [Base Sepolia explorer](https://sepolia.basescan.org/)
```
We're a little more lenient on external links in assets, but this is a bit much. Don't include external links please!

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+┌─────────────────────────────────────┐
+│  Applications (DeFi, DAO, Gaming)   │
+├─────────────────────────────────────┤
+│  Dual-Mode Tokens ([ERC-20](./eip-20) + Privacy)│  ← Optional privacy
```
Links won't work in code fences.

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+- **Merkle Tree**: Authenticated structure storing commitments for zero-knowledge membership proofs
+- **Proof Type**: Parameter routing different proof strategies
+- **View Tag**: Single-byte scanning optimization (OPTIONAL but RECOMMENDED)
+- **Stealth Address**: One-time recipient address from ephemeral keys (OPTIONAL)
```
Creating requirements in the definition section is weird. Avoid doing that.

0xRowan replied:
Fix: https://github.com/ethereum/ERCs/pull/1359/commits/ed9ee69669c69d8dd3409bfc5a09cd126e79c5ba

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+- Tree structures (single vs. dual-layer)
+- Public signals schemas
+
+To enable **client interoperability** across different implementations, this standard defines an **OPTIONAL but RECOMMENDED** Privacy Configuration File mechanism, inspired by [ERC-8004](./eip-8004)'s Agent Registration File pattern.
```
```suggestion
To enable **client interoperability** across different implementations, this standard defines an **OPTIONAL but RECOMMENDED** Privacy Configuration File mechanism, inspired by [ERC-8004](./eip-8004.md)'s Agent Registration File pattern.
```

Link to the markdown files please.

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+
+- For `bn128`: This is the BN254 scalar field prime (Fr), a 254-bit prime
+- For `bls12-381`: Use the BLS12-381 scalar field prime
+- Can be obtained from cryptographic libraries (e.g., `snarkjs`, `circomlibjs`)
```
Avoid mentioning specific libraries and products. As the proposal ages, old libraries may get abandoned and new better libraries may arise.

SamWilsn commented on ERCS/erc-8086.md:
````diff
@@ -0,0 +1,970 @@
+   - Encrypt notes using specified algorithm and parameters
+   - Compute hashes using specified hash function
+   - Interact with the token contract using correct proofType values
+```
````
This is not code.

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+- **Dual-Mode Protocols** focus on mode conversion logic (their core value)
+- **Wrapper Protocols** focus on wrapping/unwrapping logic (their core value)
+- **Privacy layer uses unified interface** (this standard)
+- Result: Faster ecosystem expansion in the privacy direction
```
This is absolutely Motivation.

0xRowan replied:
I’ve removed this, as there’s already a related description in the **Motivation** section.

SamWilsn commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+
+### Why Include Metadata Functions?
+
+`name()`, `symbol()`, `decimals()` are marked OPTIONAL but RECOMMENDED.
```
Same note as earlier. Requirements stay in the Specification section.

0xRowan replied:
I’ve removed this, as there’s already a related description in the Specification section.

SamWilsn commented on ERCS/erc-8086.md:
````diff
@@ -0,0 +1,970 @@
+- Different encryption algorithms
+- Different tree structures
+- Different proof encoding formats
+```
````
Not code.

0xRowan replied:
Fixed

SamWilsn reviewed (commented):
If I had to guess, this looks like it was mostly AI generated? That isn't a problem, but I'd ask that you go through and review it in more detail, and cut out the repetitive bits and the obvious rationale parts.

0xRowan commented on ERCS/erc-8085.md:
```diff
@@ -0,0 +1,457 @@
+   - Transparent mode satisfies compliance requirements
+   - Privacy mode available when legally permitted
+   - Can respond to changing jurisdictional rules by adjusting mode usage
+
```
I will remove this, as similar content is already covered in the Motivation section.

0xRowan committed:
Update erc-8085.md

0xRowan commented on assets/erc-8086/README.md:
```diff
@@ -0,0 +1,128 @@
+3. Get test ETH from [Base Sepolia faucet](https://www.alchemy.com/faucets/base-sepolia)
+4. Mint privacy tokens
+5. Perform privacy transfers
+6. Verify transactions on [Base Sepolia explorer](https://sepolia.basescan.org/)
```
I will removed all the links from this file.

0xRowan committed:
Remove the links in README.md

0xRowan committed:
Remove the links in README.md

0xRowan commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+---
+eip: 8086
+title: Privacy Token
+description: A minimal interface for native privacy-preserving fungible tokens on Ethereum
```
I’ll change it to: An abstract privacy foundation layer that can be used in both dual-mode tokens and wrapped tokens.

0xRowan committed:
Update erc-8086.md

0xRowan committed:
Restructure ERC-8086 Abstract for clarity and coherence

0xRowan committed:
Fix: Links won't work in code fences.

0xRowan committed:
Creating requirements in the definition section is weird. Avoid doing that.

0xRowan committed:
Link to the markdown files

0xRowan committed:
Remove all these horizontal rules

0xRowan commented on ERCS/erc-8086.md:
```diff
@@ -0,0 +1,970 @@
+
+**Client Usage**: Clients SHOULD check this field first to ensure they can parse the configuration format. Unknown types SHOULD be rejected.
+
+---
```
Fix: https://github.com/ethereum/ERCs/pull/1359/commits/3b0e0a84f63b8028c4c52dfd28f211bb94117303

0xRowan committed:
Avoid mentioning specific libraries and products

0xRowan committed:
Make sure only code is wrapped in backticks.

0xRowan committed:
Rename toPrivacy to toPrivate and remove dedicated conversion events

  - Rename `toPrivacy()` to `toPrivate()`
  - Remove `ConvertToPrivacy` and `ConvertToPublic` events (redundant with ERC-20 Transfer and ERC-8086 events)
  - Add Transfer event requirements for mode conversions in spec
  - Sync changes to reference implementation (IDualModeToken.sol, DualModeToken.sol, README.md)

SamWilsn commented on assets/erc-8086/deployments/base-sepolia.json:
Is this file necessary?

SamWilsn commented on assets/erc-8086/README.md:
The renderer doesn't automatically generate directory listings, so it can be helpful to manually link to the files from the README so people can find them.

SamWilsn reviewed (approved):
Looks fine for a draft!
